src/brod_consumer.erl

```diff
@@ -418,38 +395,63 @@ do_debug(Pid, Debug) ->
-                                 State#state.min_bytes),
-  MaxBytes = proplists:get_value(max_bytes, NewCbOptions,
-                                 State#state.max_bytes),
-  MaxWaitTime = proplists:get_value(max_wait_time, NewCbOptions,
-                                    State#state.max_wait_time),
-  SleepTimeout = proplists:get_value(sleep_timeout, NewCbOptions,
-                                     State#state.sleep_timeout),
-  PrefetchCount = proplists:get_value(prefetch_count, NewCbOptions,
-                                      State#state.prefetch_count),
-  Offset = proplists:get_value(offset, NewCbOptions,
-                               State#state.offset),
-  State1 = State#state{ offset = Offset
-                      , min_bytes = MinBytes
-                      , max_bytes = MaxBytes
-                      , max_wait_time = MaxWaitTime
-                      , sleep_timeout = SleepTimeout
-                      , prefetch_count = PrefetchCount
-                      },
-  %% in case callback wants to fetch from a new offset, rest fetch buffer
+-spec update_options(options(), #state{}) -> {ok, #state{}} | {error, any()}.
+update_options(Options, #state{begin_offset = OldBeginOffset} = State) ->
+  F = fun(Name, Default) -> proplists:get_value(Name, Options, Default) end,
   NewState =
-    case OldOffset =/= Offset of
-      true  ->
-        State1;
-      false ->
-        State1#state{ cb_pending = []
-                    , cb_pending_cnt = 0
-                    , last_corr_id = undefined
-                    }
-    end,
-  {ok, NewState}.
+    State#state{ begin_offset   = F(begin_offset, State#state.begin_offset)
+               , min_bytes      = F(min_bytes, State#state.min_bytes)
+               , max_bytes      = F(max_bytes, State#state.max_bytes)
+               , max_wait_time  = F(max_wait_time, State#state.max_wait_time)
+               , sleep_timeout  = F(sleep_timeout, State#state.sleep_timeout)
+               , prefetch_count = F(prefetch_count, State#state.prefetch_count)
+               },
+  case NewState#state.begin_offset =:= OldBeginOffset of
+    true  ->
+      {ok, NewState};
+    false ->
+      %% reset buffer in case subscriber wants to fetch from a new offset
+      NewState1 = NewState#state{pending_acks = []},
+      resolve_begin_offset(NewState1)
+  end.
+
+-spec resolve_begin_offset(#state{}) -> {ok, #state{}} | {error, any()}.
+resolve_begin_offset(#state{ begin_offset = BeginOffset
+                           , socket_pid   = SocketPid
+                           , topic        = Topic
+                           , partition    = Partition
+                           } = State) ->
+  case fetch_valid_offset(SocketPid, BeginOffset, Topic, Partition) of
+    {ok, NewBeginOffset} ->
+      {ok, State#state{begin_offset = NewBeginOffset}};
+    {error, Reason} ->
+      {error, Reason}
+  end.
+
+fetch_valid_offset(SocketPid, InitialOffset, Topic, Partition) ->
```

Comment:
Kafka does not understand -1 and -2 in fetch requests, you first need to resolve this to a valid number via offset request.